code/datums/elements/movetype_handler.dm

```diff
@@ -0,0 +1,110 @@
+	RegisterSignal(AM, COMSIG_PAUSE_FLOATING_ANIM, .proc/pause_floating_anim)
+	attached_atoms[AM] = TRUE
+
+	if(AM.movement_type & (FLOATING|FLYING) && !HAS_TRAIT(AM, TRAIT_NO_FLOATING_ANIM))
+		float(AM)
+
+/datum/element/movetype_handler/Detach(datum/source)
+	UnregisterSignal(source, list(
+		GLOB.movement_type_trait_add_signals,
+		GLOB.movement_type_trait_remove_signals,
+		SIGNAL_ADDTRAIT(TRAIT_NO_FLOATING_ANIM),
+		SIGNAL_REMOVETRAIT(TRAIT_NO_FLOATING_ANIM),
+		COMSIG_PAUSE_FLOATING_ANIM
+	))
+	attached_atoms -= source
+	paused_floating_anim_atoms -= source
+	stop_floating(source)
+	return ..()
+
+/// Called when a movement type trait is added to the movable. Enables the relative bitflag.
+/datum/element/movetype_handler/proc/on_movement_type_trait_gain(atom/movable/source, trait)
+	SIGNAL_HANDLER
+	var/flag = GLOB.movement_type_trait_to_flag[trait]
+	if(source.movement_type & flag)
+		return
+	if(!(source.movement_type & (FLOATING|FLYING)) && (trait == TRAIT_MOVE_FLYING || trait == TRAIT_MOVE_FLOATING) && !paused_floating_anim_atoms[source] && !HAS_TRAIT(source, TRAIT_NO_FLOATING_ANIM))
+		float(source)
+	source.movement_type |= flag
+	SEND_SIGNAL(source, COMSIG_MOVETYPE_FLAG_ENABLED, flag)
+
+/// Called when a movement type trait is removed from the movable. Disables the relative bitflag if it wasn't there in the compile-time bitfield.
+/datum/element/movetype_handler/proc/on_movement_type_trait_loss(atom/movable/source, trait)
+	SIGNAL_HANDLER
+	var/flag = GLOB.movement_type_trait_to_flag[trait]
+	if(initial(source.movement_type) & flag)
+		return
+	source.movement_type &= ~flag
+	if((trait == TRAIT_MOVE_FLYING || trait == TRAIT_MOVE_FLOATING) && !(source.movement_type & (FLOATING|FLYING)))
+		stop_floating(source)
+	SEND_SIGNAL(source, COMSIG_MOVETYPE_FLAG_DISABLED, flag)
+
+/// Called when the TRAIT_NO_FLOATING_ANIM trait is added to the movable. Stops it from bobbing up and down.
+/datum/element/movetype_handler/proc/on_no_floating_anim_trait_gain(atom/movable/source, trait)
+	SIGNAL_HANDLER
+	stop_floating(source)
+
+/// Called when the TRAIT_NO_FLOATING_ANIM trait is removed from the mob. Restarts the bobbing animation.
+/datum/element/movetype_handler/proc/on_no_floating_anim_trait_loss(atom/movable/source, trait)
+	SIGNAL_HANDLER
+	if(source.movement_type & (FLOATING|FLYING) && !paused_floating_anim_atoms[source])
+		float(source)
+
+///Pauses the floating animation for the duration of the timer... plus [tickrate - (world.time + timer) % tickrate] to be precise.
+/datum/element/movetype_handler/proc/pause_floating_anim(atom/movable/source, timer)
+	SIGNAL_HANDLER
+	if(paused_floating_anim_atoms[source] < world.time + timer)
+		stop_floating(source)
+		if(!length(paused_floating_anim_atoms))
+			START_PROCESSING(SSdcs, src) //1 second tickrate.
+		paused_floating_anim_atoms[source] = world.time + timer
+
+/datum/element/movetype_handler/process()
+	for(var/a in paused_floating_anim_atoms)
+		var/atom/movable/AM = a
```

Comment:
Better variable names. Something like this is good:

```
	for(var/_floating in paused_floating_anim_atoms)
		var/atom/movable/floating = _floating
```

Reply:
I don't want to comment on all of them specifically, but just change every variable you create that's `AM`, `L`, `M`, etc too be more exact to what they actually are.